torch/csrc/autograd/python_engine.cpp

```diff
@@ -392,8 +392,23 @@ PyObject* THPEngine_queue_callback(PyObject* self, PyObject* _callback) {
   engine.queue_callback([callback]() {
     pybind11::gil_scoped_acquire gil;
     THPObjectPtr result{PyObject_CallFunctionObjArgs(callback.get(), nullptr)};
-    if (!result)
-      throw python_error();
+    if (!result) {
+      // Note [ Persisting PyErr state across autograd engine threads ]
+      //
+      // Throw a python_error with the PyErr state persisted so that we don't
+      // lose the error state when returning to the cpu thread. Ordinarily, if
+      // we were calling into python during backward, e.g. executing a hook
+      // or custom autograd Function, we would be able to directly use
+      // `throw python_error()`. This is because (1) the engine has logic around
+      // evaluate_function that catches and persists errors, and (2) we init the
+      // PyThreadState on the creation of the thread so we aren't actually
+      // creating a separate PyThreadState everytime we call gil_scoped_acquire.
+      // Errors raised by queue_callback won't be caught by the try-catch around
+      // evaluate_function, so we need to replicate the persisting logic here.
```

Comment:
I'm not sure about this comment.
I think we should make this note a big more general and add it to the error handling in evaluate function as well.
It should then say that:
- Since the engine is using multithreading, it must preserve the python error from the worker thread and rethrow it as-is in the calling thread.
- This is done by persisting errors in the two places where python errors can happen: evaluate function and callback.


I'm not sure what the point related to (2) is so I would suggest removing it.

Reply:
That's a good idea. I will update this note.

(2) is the reason we shouldn't need to immediately persist using something like `throw_python_error()`

Reply:
I haven't verified (2) myself, but per the comment in `PythonEngine::thread_init`
```
// Create a PyThreadState, but release the GIL. This lets
// pybind11::gil_scoped_acquire calls inside thread_main acquire the GIL
// without having to create a new PyThreadState each time.
```

Reply:
Sure but that comment is just about speeding up gil acquisition by not having to re-allocate the state everytime. It doens't have anything to do with error handling AFAIK

Reply:
Since the PyError indicators are stored on the state, I think that would go away if we need to re-allocated it. More context here as well https://github.com/pytorch/pytorch/pull/34845